adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -287,47 +261,175 @@ def _final_callback(self):
             self._update_avg('sqr_avg', grad_sqr, theta)
             self._update_avg('var_avg', grad_var, theta)
 
-    def step(self, *args, **kwargs):
+    def zero_grad(self, *args, **kwargs):
```

Comment:
I think the logic in `zero_grad`, `patch_optimizer`, and `run_step` should belong in `ScalingRuleBase` rather than here.